contracts/collection.scilla

```diff
@@ -0,0 +1,306 @@
+field requests: Map ByStr20 (Map Uint256 (Map ByStr20 TokenState)) = Emp ByStr20 (Map Uint256 (Map ByStr20 TokenState))
+
+(* Mapping from collection_id to collection_fee *)
+field collection_commission_fee: Map Uint32 Uint128 = Emp Uint32 Uint128
+
+(* Mapping for knowing if an NFT has been sold or not *)
+field has_had_first_sale: Map ByStr20 (Map Uint256 Bool) = Emp ByStr20 (Map Uint256 Bool)
+
+
+(* Emit Errors *)
+procedure Throw(error: Error)
+  e = make_error error;
+  throw e
+end
+
+procedure RequireTokenOwner(
+    token_address: ByStr20 with contract 
+        field token_owners: Map Uint256 ByStr20 
+    end,
+    token_id: Uint256
+    )
+    owner <- & token_address.token_owners[token_id];
+
+    match owner with
+        | Some owner =>
+        owner_is_sender = builtin eq owner _sender;
+            match owner_is_sender with
+            | True => (* No operation *)
+            | False =>
+                error = SenderIsNotTokenOwner;
+                Throw error
+            end
+        | None =>
+        error = TokenIdDoesNotExist;
+        Throw error
+    end
+end
+
+procedure RequireTokenNotInCollection(
+    token_address: ByStr20 with contract 
+        field token_owners: Map Uint256 ByStr20 
+    end,
+    token_id: Uint256
+    )
+    is_token_in_collection <- exists token_collection[token_address][token_id];
+
+    match is_token_in_collection with 
+    | True => 
+        (* A token can ONLY belong to ONE collection - TokenAlreadyInCollection *)
+        error = TokenAlreadyInCollection;
+        Throw error
+    | False => (* No Operation *)
+    end 
+end
+
+procedure RequireBrandOwner(
+    token_address: ByStr20 with contract 
+        field token_owners: Map Uint256 ByStr20 
+    end,
+    token_id: Uint256
+    )
+    owner <- & token_address.token_owners[token_id];
+
+    match owner with
+        | Some owner =>
+        owner_is_sender = builtin eq owner _sender;
+            match owner_is_sender with
+            | True => (* No operation *)
+            | False =>
+                error = SenderIsNotTokenOwner;
+                Throw error
+            end
+        | None =>
+        error = TokenIdDoesNotExist;
+        Throw error
+    end
+end
+
+(* We allow a max upper boundary of 250 basis points (2.5%) *)
+(* Any more than that will result in a failed transaction *)
+procedure RequireValidBPS(val_bps: Uint128)
+    is_lte_max = uint128_le val_bps max_commission_fee_bps;
+    match is_lte_max with
+        | True => (* No Operation *)
+        | False =>
+        error = CommissionFeeTooHigh;
+        Throw error
+    end
+end
+
+
+(* Lets anyone create an empty collection with an id assigned by the smart contract *)
+(* @dev should integrate with allowlist to only allow whitelisted addresses *)
+transition CreateCollection(
+    commission_fee: Uint128
+    )
+    RequireValidBPS commission_fee;
+    RequireValidBrand; (* Placeholder for requiring only whitelisted brands *)
```

Comment:
@sam-zilliqa can we have `RequireContractOwner` here instead of `RequireValidBrand`? This would ensure that collections can only be created within Rialto (i.e. by the Rialto backend). This would simplify the synchronization of onchain/offchain data (by both Indexer and Rialto). 
You can refer to the [allowlist](https://github.com/Zilliqa/marketplace-contracts/blob/605aedbb8fe9fd9e56fb87999094105b95b4b312/contracts/allowlist.scilla#L157) implementation.

Reply:
Hey I'm a bit confused by your comment - [I just did a commit](https://github.com/Zilliqa/marketplace-contracts/pull/54/commits/6f6d0ffebcdd5c48e08d264328901b7633d64e87) where you can see my idea of this. I simply looked at how allowlist was implemented for the marketplace contracts and added the same for `collection.scilla` (was a quick edit). 

Now all "user" transitions have an additional require, where the relevant address is required to be on the allowlist. The allowlist references the same allowlist contract used by the marketplace contracts. 

Let me know if I misunderstood something

Reply:
Hi @StayFoolisj, we had this discussion with the team earlier about using `RequireContractOwner` to ensure that collections are created on the Rialto platform (i.e. the Rialto Backend, which has the key for signing the tx, would call this transition on behalf of a brand -- same way a wallet gets added to the allowlist contract). That is, this check would prevent a collection from being created outside Rialto. We wanted to enforce this because collection metadata are captured off-chain, during collection creation call on the Rialto platform. If this was called outside Rialto, the Rialto backend will not be able to save collection metadata on its database. 

However, this may not be necessary since we can accept that this scenario is outside of scope and won't be supported. That is, if a collection is created outside of Rialto, then this collection won't be shown on the Rialto app.

Wdyt @philippeatzilliqa @lukozill

Reply:
@sam-zilliqa we can stick with `RequireAllowedUser` check that you have already implemented.